src/sasl/cyrsasl_scram.erl

```diff
@@ -58,164 +58,194 @@ mech_new(_Host, Creds) ->
-                                           if CompareStoredKey == State#state.stored_key ->
-                                                  ServerSignature =
-                                                  mongoose_scram:server_signature(State#state.server_key,
-                                                                                  AuthMessage),
-                                                  R = [{username, State#state.username},
-                                                       {sasl_success_response,
-                                                        <<"v=", (jlib:encode_base64(ServerSignature))/binary>>},
-                                                       {auth_module, State#state.username}],
-                                                  {ok, mongoose_credentials:extend(State#state.creds, R)};
-                                              true -> {error, <<"bad-auth">>}
-                                           end;
-                                       _Else -> {error, <<"bad-protocol">>}
-                                   end;
-                               {$r, _} -> {error, <<"bad-nonce">>};
-                               _Else -> {error, <<"bad-protocol">>}
-                           end;
-                       true -> {error, <<"bad-channel-binding">>}
-                    end;
-                _Else -> {error, <<"bad-protocol">>}
+    ClientInList = re:split(ClientIn, <<",">>),
+    ClientInSteps = [ fun parse_step4_client_in/1,
+                      fun parse_channel_binding_attribute/1],
+    case parse_attributes(ClientInList, ClientInSteps) of
+        {ok, {NonceAtt, ClientProofAtt}} ->
+            Nonce = <<(State#state.client_nonce)/binary,
+                      (State#state.server_nonce)/binary>>,
+            {PStart, _} = binary:match(ClientIn, <<",p=">>),
+            AuthMessage = iolist_to_binary( [State#state.auth_message,",",
+                                    binary:part(ClientIn, 0, PStart)]),
+            AuthSteps = [fun verify_nonce/1,
+                         fun get_client_proof/1,
+                         fun get_stored_key/1,
+                         fun verify_stored_key/1],
+            AuthArgs = [{NonceAtt, Nonce}, ClientProofAtt, AuthMessage, State],
+            case parse_attributes(AuthArgs, AuthSteps) of
+                ok ->
+                    ServerSignature =
+                        mongoose_scram:server_signature(State#state.server_key,
+                                                        AuthMessage),
+                    R = [{username, State#state.username}, {sasl_success_response,
+                         <<"v=", (jlib:encode_base64(ServerSignature))/binary>>},
+                         {auth_module, State#state.username}],
+                    {ok, mongoose_credentials:extend(State#state.creds, R)};
+                {error, Reason} ->
+                    {error, Reason}
```

Comment:
Why don't we extract this part to another function as well? That's already two nested cases, not too much but still, not flat 🤔 
Similar idea in the step 2, what do you think?

Reply:
I wanted to separate validation of client input and authentication verification. The problem with using single fold (and avoiding the second case statement) would be that we'd need to carry all the arguments though all the functions. I think that with what we have now is a balance between nesting and code readability. I wanted to leave as much core functionality in mech_step as possible. Reading mech_step alone we could have some basic understanding to what is being done without going though all the functions.